plugins/inputs/neptune_apex/neptune_apex.go

```diff
@@ -0,0 +1,265 @@
+type NeptuneApex struct {
+	Servers         []string
+	ResponseTimeout internal.Duration
+	httpClient      *http.Client
+}
+
+// Description implements telegraf.Input.Description
+func (*NeptuneApex) Description() string {
+	return "Neptune Apex data collector"
+}
+
+// SampleConfig implements telegraf.Input.SampleConfig
+func (*NeptuneApex) SampleConfig() string {
+	return `
+  ## The Neptune Apex plugin reads the publicly available status.xml data from a local Apex.
+  ## Measurements will be logged under "apex".
+
+  ## The hostname/IP of the local Apex(es). If you specify more than one server, they will
+  ## be differentiated by the "hostname" tag.
+  servers = [
+    "apex.local",
+  ]
+
+  ## The response_timeout specifies how long to wait for a reply from the Apex.
+  #response_timeout = "5s"
+`
+}
+
+// Gather implements telegraf.Input.Gather
+func (n *NeptuneApex) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+	for _, server := range n.Servers {
+		wg.Add(1)
+		go func(server string) {
+			defer wg.Done()
+			acc.AddError(n.gatherServer(acc, server))
+		}(server)
+	}
+	wg.Wait()
+	return nil
+}
+
+func (n *NeptuneApex) gatherServer(acc telegraf.Accumulator, server string) error {
+	resp, err := n.sendRequest(server)
+	if err != nil {
+		return err
+	}
+	return n.parseXML(acc, resp)
+}
+
+// parseXML is strict on the input and does not do best-effort parsing. This is because of the life-support nature
+// of the Neptune Apex.
+func (n *NeptuneApex) parseXML(acc telegraf.Accumulator, data []byte) error {
+	r := xmlReply{}
+	err := xml.Unmarshal(data, &r)
+	if err != nil {
+		return fmt.Errorf("unable to unmarshal XML: %v\nXML DATA: %q", err, data)
+	}
+
+	var reportTime time.Time
+	if reportTime, err = parseTime(r.Date, r.Timezone); err != nil {
+		return err
+	}
+
+	mainFields := map[string]interface{}{
+		"software":       r.SoftwareVersion,
+		"hardware":       r.HardwareVersion,
+		"serial":         r.Serial,
+		"timezone":       r.Timezone,
+		"power_failed":   r.PowerFailed,
+		"power_restored": r.PowerRestored,
+	}
+	acc.AddFields(Measurement, mainFields, map[string]string{"hostname": r.Hostname, "type": "controller"}, reportTime)
+
+	// Outlets.
+	for _, o := range r.Outlet {
+		tags := map[string]string{
+			"hostname":  r.Hostname,
+			"output_id": o.OutputID,
+			"device_id": o.DeviceID,
+			"name":      o.Name,
+			"type":      "output",
+		}
+		fields := map[string]interface{}{
+			"state": o.State,
+		}
+		// Find Amp and Watt probes and add them as fields. Remove the redundant probe.
+		if pos := findProbe(fmt.Sprintf("%sW", o.Name), r.Probe); pos > -1 {
+			value, err := strconv.ParseFloat(strings.TrimSpace(r.Probe[pos].Value), 64)
+			if err != nil {
+				acc.AddError(
+					fmt.Errorf("cannot convert string value %q to float64: %v", r.Probe[pos].Value, err))
+				continue // Skip the whole outlet.
+			}
+			fields["watt"] = value
+			r.Probe[pos] = r.Probe[len(r.Probe)-1]
```

Comment:
This could panic if r.Probe length is 0, is this possible such in the case of unexpected XML?

Reply:
That statement is part of an if block that guards against r.Probe from having 0 elements.